keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,458 @@
+
+There are no metrics that would inform anyone that the feature was
+failing (discussed more below).
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Assuming the "Manual" model, they could check all services to see if
+they were annotated to use the feature. Assuming the "Automatic"
+model, there is no easy way to determine thisl
+
+###### How can someone using this feature know that it is working for their instance?
+
+They can know that it hasn't broken anything because their services
+won't be broken.
+
+There's no simple way to confirm that kube-proxy is optimizing traffic
+correctly based on this feature. The user would need to packet-sniff,
+or check individual endpoint logs, or look at the iptables rules
+generated by kube-proxy.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+Services using node-level topology should be as reliable as services
+_not_ using node-level topology, and should have lower latency.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [X] Other
+
+There are no existing metrics for determining the availability or
+response time of individual Services, and this KEP does not add any.
+In general, the operator cannot know what response times the user
+expects from their Services anyway.
+
+The developer may be able to generate their own metrics based on logs
+from the endpoint pods and/or logs generated by the service's clients;
+but this involves user-owned code, not kubernetes code, on both ends.
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+Not really. We could expose the number of node-local-topology services
+on each node, but it's not clear that that would be useful.
+
+### Dependencies
+
+###### Does this feature depend on any specific services running in the cluster?
+
+It depends on a service proxy which understands the feature. We will
+update `kube-proxy` ourselves, but network plugins / kubernetes
+distributions that ship their own alternative service proxies will
+also need to be updated to support the feature before their users can
+benefit from it.
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+
+No
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+It might add a new field to EndpointSliceHints (which would be unset
+in most EndpointSlices).
```

Comment:
The more I think about this, the more I think there's no real value in adding a new hints field since we already have the nodeName for each endpoint. If we were going to try to be clever and assign endpoints proportionally to nodes (a la zonal hints), maybe we should add a field.

Reply:
If we have "automatic" node-level topology then the new Hint is useful because it lets you see when we decided to use node-level topology for a service. But if we stick with "manual" then you don't need that.